core/src/main/java/com/linecorp/armeria/common/stream/TimeoutStreamMessage.java

```diff
@@ -0,0 +1,270 @@
+     *
+     * @param cause the cause of the abort
+     */
+    @Override
+    public void abort(Throwable cause) {
+        cancelSchedule();
+        delegate.abort(cause);
+    }
+
+    static final class TimeoutSubscriber<T> implements Runnable, Subscriber<T>, Subscription {
+
+        private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
+        private final Subscriber<? super T> delegate;
+        private final EventExecutor executor;
+        private final StreamTimeoutMode timeoutMode;
+        private final Duration timeoutDuration;
+        private final long timeoutNanos;
+        @Nullable
+        private ScheduledFuture<?> timeoutFuture;
+        @Nullable
+        private Subscription subscription;
+        private long lastEventTimeNanos;
+        private boolean completed;
+
+        TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration,
+                          StreamTimeoutMode timeoutMode) {
+            this.delegate = requireNonNull(delegate, "delegate");
+            this.executor = requireNonNull(executor, "executor");
+            this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+            timeoutNanos = timeoutDuration.toNanos();
+            this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+        }
+
+        private ScheduledFuture<?> scheduleTimeout(long delay, TimeUnit unit) {
+            return executor.schedule(this, delay, unit);
+        }
+
+        void cancelSchedule() {
+            if (timeoutFuture != null && !timeoutFuture.isCancelled()) {
+                timeoutFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public void run() {
+            if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+                final long currentTimeNanos = System.nanoTime();
+                final long elapsedNanos = currentTimeNanos - lastEventTimeNanos;
+
+                if (elapsedNanos < timeoutNanos) {
+                    final long delayNanos = timeoutNanos - elapsedNanos;
+                    timeoutFuture = scheduleTimeout(delayNanos, TimeUnit.NANOSECONDS);
+                    return;
+                }
+            }
+            completed = true;
+            delegate.onError(new TimeoutException(
+                    String.format(TIMEOUT_MESSAGE, timeoutDuration.toMillis(), timeoutMode)));
+            subscription.cancel();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            subscription = s;
+            delegate.onSubscribe(this);
+            if (completed) {
+                return;
+            }
+            lastEventTimeNanos = System.nanoTime();
+            timeoutFuture = scheduleTimeout(timeoutNanos, TimeUnit.NANOSECONDS);
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (completed) {
+                return;
```

Comment:
to avoid potential leaks

```suggestion
            if (completed) {
                PooledObjects.close(t);
                return;
```